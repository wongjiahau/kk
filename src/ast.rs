use crate::non_empty::NonEmpty;
/// The syntax tree here represents raw syntax tree that is not type checked

#[derive(Debug, Clone)]
pub enum Statement {
    Let {
        keyword_let: Token,
        left: Token,
        type_variables: Vec<Token>,
        right: Expression,
        type_annotation: Option<TypeAnnotation>,
    },
    Type {
        keyword_type: Token,
        left: Token,
        right: TypeAnnotation,
        type_variables: Vec<Token>,
    },
    Enum {
        keyword_enum: Token,
        name: Token,
        constructors: Vec<EnumConstructor>,
        type_variables: Vec<Token>,
    },
    Do {
        keyword_do: Token,
        expression: Expression,
    },
}

#[derive(Debug, Clone)]
pub struct EnumConstructor {
    pub name: Token,
    pub left_parenthesis: Token,
    pub payload: Option<Box<TypeAnnotation>>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct TypeVariable {
    pub token: Token,
}

#[derive(Debug, Clone)]
pub enum Type {
    Underscore,

    /// Type variable that is declared. Cannot be substituted before instantiation.
    /// Note that a type variable is only explicit within its own scope
    ExplicitTypeVariable {
        name: String,
    },

    /// Type variable that is implicitly created  for unification.
    ImplicitTypeVariable {
        name: String,
    },
    Record {
        key_type_pairs: Vec<(String, Type)>,
    },
    Named {
        name: String,
        type_arguments: Vec<(String, Type)>,
    },
    Function(FunctionType),
    Tuple(Box<NonEmpty<Type>>),
    Boolean,
    Float,
    Integer,
    String,
    Character,
    Null,
    Array(Box<Type>),
}

#[derive(Debug, Clone)]
pub struct TagType {
    pub tagname: String,
    pub payload: Option<Box<Type>>,
}

#[derive(Debug, Clone)]
pub struct TypeScheme {
    pub type_variables: Vec<String>,
    pub type_value: Type,
}

#[derive(Debug, Clone)]
pub struct FunctionType {
    pub parameters_types: Box<NonEmpty<Type>>,
    pub return_type: Box<Type>,
}

#[derive(Debug, Clone)]
pub enum SymbolSource {
    BuiltIn,
    UserDefined {
        filename: String,
        position: Position,
    },
}

#[derive(Debug, Clone)]
pub struct NamedTypeAnnotationArguments {
    pub left_angular_bracket: Token,
    pub arguments: Vec<TypeAnnotation>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub enum TypeAnnotation {
    Named {
        name: Token,
        type_arguments: Option<TypeArguments>,
    },
    Record {
        left_curly_bracket: Token,
        key_type_annotation_pairs: Vec<(Token, TypeAnnotation)>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        element_type: Box<TypeAnnotation>,
        right_square_bracket: Token,
    },
    Underscore(Token),
    Function {
        start_token: Token,
        parameters_types: Box<NonEmpty<TypeAnnotation>>,
        return_type: Box<TypeAnnotation>,
    },
}

#[derive(Debug, Clone)]
pub enum DestructurePattern {
    String(Token),
    Character(Token),
    Integer(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Null(Token),
    Underscore(Token),
    Identifier(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<DestructurePattern>>,
        right_parenthesis: Token,
    },
    Record {
        left_curly_bracket: Token,
        key_value_pairs: Vec<DestructuredRecordKeyValue>,
        right_curly_bracket: Token,
    },
    Array {
        left_square_bracket: Token,
        right_square_bracket: Token,
        spread: Option<DestructurePatternArraySpread>,
    },
    Tuple(DestructurePatternTuple),
}
#[derive(Debug, Clone)]
pub struct DestructurePatternTuple {
    pub parentheses: Option<(/*left*/ Token, /*right*/ Token)>,
    pub values: Box<NonEmpty<DestructurePattern>>,
}

#[derive(Debug, Clone)]
pub struct DestructurePatternArraySpread {
    pub first_element: Box<DestructurePattern>,
    pub spread_token: Token,
    pub rest_elements: Box<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub struct DestructuredRecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub as_value: Option<DestructurePattern>,
}

#[derive(Debug, Clone)]
pub enum Expression {
    Null(Token),
    Boolean {
        token: Token,
        value: bool,
    },
    Float(Token),
    Integer(Token),
    String(Token),
    Character(Token),
    Variable(Token),
    EnumConstructor {
        scoped_name: ScopedName,
        left_parenthesis: Token,
        payload: Option<Box<Expression>>,
        right_parenthesis: Token,
    },
    Function(Box<Function>),
    FunctionCall(Box<FunctionCall>),
    Record {
        left_curly_bracket: Token,
        spread: Option<Box<Expression>>,
        key_value_pairs: Vec<RecordKeyValue>,
        right_curly_bracket: Token,
    },
    RecordAccess {
        expression: Box<Expression>,
        property_name: Token,
    },
    Array {
        left_square_bracket: Token,
        elements: Vec<Expression>,
        right_square_bracket: Token,
    },
    Let {
        keyword_let: Token,
        left: Box<DestructurePattern>,
        type_annotation: Option<TypeAnnotation>,
        right: Box<Expression>,
        true_branch: Box<Expression>,
        false_branch: Option<Box<Function>>,
    },
}

#[derive(Debug, Clone)]
pub struct RecordKeyValue {
    pub key: Token,
    pub type_annotation: Option<TypeAnnotation>,
    pub value: Expression,
}

#[derive(Debug, Clone)]
pub struct FunctionCall {
    pub function_name: Token,
    pub first_argument: Box<Expression>,
    pub rest_arguments: Option<FunctionCallRestArguments>,
    pub type_arguments: Option<TypeArguments>,
}

#[derive(Debug, Clone)]
pub struct TypeArguments {
    pub left_angular_bracket: Token,
    pub substitutions: Vec<(/*type_variable_name*/ Token, TypeAnnotation)>,
    pub right_angular_bracket: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionCallRestArguments {
    pub left_parenthesis: Token,
    pub arguments: Vec<Expression>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct Function {
    pub branches: NonEmpty<FunctionBranch>,
}

#[derive(Debug, Clone)]
pub struct FunctionBranch {
    pub start_token: Token,
    pub parameters: FunctionParameters,
    pub body: Box<Expression>,
    pub return_type_annotation: Option<TypeAnnotation>,
}

impl FunctionBranch {
    pub fn parameters(&self) -> NonEmpty<FunctionParameter> {
        match &self.parameters {
            FunctionParameters::NoParenthesis { parameter } => NonEmpty {
                head: parameter.clone(),
                tail: vec![],
            },
            FunctionParameters::WithParenthesis { parameters, .. } => parameters.clone(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum FunctionParameters {
    NoParenthesis {
        parameter: FunctionParameter,
    },
    WithParenthesis {
        left_parenthesis: Token,
        parameters: NonEmpty<FunctionParameter>,
        right_parenthesis: Token,
    },
}

#[derive(Debug, Clone)]
pub struct FunctionBranchRestArguments {
    pub left_parenthesis: Token,
    pub rest_arguments: Vec<FunctionParameter>,
    pub right_parenthesis: Token,
}

#[derive(Debug, Clone)]
pub struct FunctionParameter {
    pub destructure_pattern: DestructurePattern,
    pub type_annotation: Option<TypeAnnotation>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub position: Position,
    pub representation: String,
}

#[derive(Debug, Clone)]
pub struct ScopedName {
    pub namespaces: Vec<Token>,
    pub name: Token,
}

impl Token {
    pub fn dummy_identifier(representation: String) -> Token {
        Token {
            token_type: TokenType::Identifier,
            position: Position {
                line_start: 0,
                line_end: 0,
                character_index_start: 0,
                character_index_end: 0,
                column_start: 0,
                column_end: 0,
            },
            representation,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenType {
    KeywordLet,
    KeywordType,
    KeywordEnum,
    KeywordDo,
    KeywordElse,
    KeywordNull,
    KeywordTrue,
    KeywordFalse,
    Whitespace,
    LeftCurlyBracket,
    RightCurlyBracket,
    LeftParenthesis,
    RightParenthesis,
    LeftSquareBracket,
    RightSquareBracket,
    Newline,
    Colon,
    LessThan,
    MoreThan,
    Equals,
    Period,
    Spread,
    Comma,
    Minus,
    FatArrowRight,
    ThinArrowRight,
    Backslash,
    Underscore,
    Identifier,
    String,
    Character,
    Integer,
    Float,
    ScopeResolution,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Source {
    File { path: String },
    NonFile { env_name: String },
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct Position {
    pub line_start: usize,
    pub line_end: usize,
    pub column_start: usize,
    pub column_end: usize,
    pub character_index_start: usize,
    pub character_index_end: usize,
}
