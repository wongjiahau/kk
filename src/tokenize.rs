use crate::ast::*;
use peeking_take_while::PeekableExt;

#[derive(Debug, Clone)]
pub struct Character {
    index: usize,
    line_number: usize,
    column_number: usize,
    value: char,
}

pub fn tokenize(input: String) -> Result<Vec<Token>, ParseError> {
    let characters: Vec<Character> = input
        .split('\n')
        .enumerate()
        .flat_map(|(line_number, line)| {
            let mut characters = line.chars().into_iter().collect::<Vec<char>>();
            characters.push('\n');
            characters
                .into_iter()
                .enumerate()
                .map(move |(column_number, character)| (character, line_number, column_number))
                .collect::<Vec<(char, usize, usize)>>()
        })
        .enumerate()
        .map(
            |(index, (character, line_number, column_number))| Character {
                index,
                value: character,
                line_number,
                column_number,
            },
        )
        .collect::<Vec<Character>>();

    let mut it = characters.into_iter().peekable();
    let mut tokens = Vec::<Token>::new();

    while let Some(character) = it.next() {
        match character.value {
            '_' => tokens.push(Token {
                token_type: TokenType::Underscore,
                representation: "_".to_string(),
                position: make_position(character, None),
            }),
            '#' => {
                let tagname: Vec<Character> = it
                    .by_ref()
                    .peeking_take_while(|charcater| {
                        charcater.value.is_alphanumeric() || charcater.value == '_'
                    })
                    .collect();

                tokens.push(Token {
                    token_type: TokenType::Tag,
                    representation: "#".to_string() + &stringify(tagname.clone()),
                    position: make_position(character, tagname.last()),
                })
            }
            '\'' | '"' => {
                let quote = character.value;
                let content: Vec<Character> = it
                    .by_ref()
                    .peeking_take_while(|character| character.value != quote)
                    .collect();

                match it.by_ref().next() {
                    Some(ending_quote) => tokens.push(Token {
                        token_type: TokenType::String,
                        representation: format!("\"{}\"", stringify(content.clone())),
                        position: make_position(character, Some(&ending_quote)),
                    }),
                    None => {
                        return Err(ParseError::UnterminatedString {
                            position: make_position(character, content.last()),
                        })
                    }
                }
            }
            '-' | '0'..='9' => {
                let intergral = it
                    .by_ref()
                    .peeking_take_while(|character| character.value.is_digit(10))
                    .collect::<Vec<Character>>();

                let fractional = match it.by_ref().peek() {
                    Some(Character { value: '.', .. }) => {
                        let _ = it.next();
                        let characters = it
                            .by_ref()
                            .peeking_take_while(|character| character.value.is_digit(10))
                            .collect::<Vec<Character>>();
                        Some(characters)
                    }
                    _ => None,
                };
                tokens.push(Token {
                    token_type: TokenType::Number,
                    representation: match &fractional {
                        Some(fractional) => format!(
                            "{}{}.{}",
                            character.value,
                            stringify(intergral.clone()),
                            stringify(fractional.clone())
                        ),
                        None => format!("{}{}", character.value, stringify(intergral.clone())),
                    },
                    position: make_position(
                        character,
                        match &fractional {
                            Some(fractional) => fractional.last(),
                            None => intergral.last(),
                        },
                    ),
                })
            }
            'A'..='Z' | 'a'..='z' => {
                let characters = it
                    .by_ref()
                    .peeking_take_while(|character| {
                        character.value.is_alphanumeric() || character.value == '_'
                    })
                    .collect::<Vec<Character>>();

                let representation =
                    format!("{}{}", character.value, stringify(characters.clone()));
                tokens.push(Token {
                    token_type: get_token_type(representation.clone()),
                    representation,
                    position: make_position(character, characters.last()),
                })
            }
            '.' => {
                let dots = it
                    .by_ref()
                    .peeking_take_while(|character| character.value == '.')
                    .collect::<Vec<Character>>();

                match dots.len() {
                    0 => tokens.push(Token {
                        token_type: TokenType::Period,
                        representation: ".".to_string(),
                        position: make_position(character, None),
                    }),
                    2 => tokens.push(Token {
                        token_type: TokenType::Spread,
                        representation: "...".to_string(),
                        position: make_position(character, dots.last()),
                    }),
                    _ => {
                        return Err(ParseError::InvalidChar {
                            error: "Only one dot (.) or three dots (...) is acceptable".to_string(),
                            position: make_position(character, dots.last()),
                        })
                    }
                }
            }
            '=' => match it.peek() {
                Some(Character { value: '>', .. }) => {
                    let greater_than_character = it.by_ref().next();
                    tokens.push(Token {
                        token_type: TokenType::ArrowRight,
                        representation: "=>".to_string(),
                        position: make_position(character, greater_than_character.as_ref()),
                    })
                }
                _ => tokens.push(Token {
                    token_type: TokenType::Equals,
                    representation: "=".to_string(),
                    position: make_position(character, None),
                }),
            },
            other => {
                tokens.push(Token {
                    position: make_position(character, None),
                    representation: other.to_string(),
                    token_type: match other {
                        '{' => TokenType::LeftCurlyBracket,
                        '}' => TokenType::RightCurlyBracket,
                        '(' => TokenType::LeftParenthesis,
                        ')' => TokenType::RightParenthesis,
                        '[' => TokenType::LeftSquareBracket,
                        ']' => TokenType::RightSquareBracket,
                        ' ' => TokenType::Whitespace,
                        ':' => TokenType::Colon,
                        ';' => TokenType::Semicolon,
                        '+' => TokenType::Plus,
                        '-' => TokenType::Minus,
                        ',' => TokenType::Comma,
                        '<' => TokenType::LessThan,
                        '>' => TokenType::MoreThan,
                        '|' => TokenType::Pipe,
                        '\\' => TokenType::Backslash,
                        '_' => TokenType::Underscore,
                        '\n' => TokenType::Newline,
                        _ => panic!(),
                    },
                });
            }
        }
    }
    Ok(tokens)
}

pub fn make_position(first: Character, last: Option<&Character>) -> Position {
    let last = match last {
        Some(character) => character,
        None => &first,
    };
    Position {
        column_start: first.column_number,
        column_end: last.column_number,
        line_start: first.line_number,
        line_end: last.line_number,
        character_index_start: first.index,
        character_index_end: last.index,
    }
}

pub fn stringify(characters: Vec<Character>) -> String {
    characters
        .into_iter()
        .map(|character| character.value.to_string())
        .collect::<Vec<String>>()
        .join("")
}

pub fn get_token_type(s: String) -> TokenType {
    if s.eq("let") {
        TokenType::KeywordLet
    } else if s.eq("type") {
        TokenType::KeywordType
    } else if s.eq("enum") {
        TokenType::KeywordEnum
    } else if s.eq("do") {
        TokenType::KeywordDo
    } else if s.eq("else") {
        TokenType::KeywordElse
    } else if s.eq("true") {
        TokenType::KeywordTrue
    } else if s.eq("false") {
        TokenType::KeywordFalse
    } else if s.eq("null") {
        TokenType::KeywordNull
    } else {
        TokenType::Identifier
    }
}
