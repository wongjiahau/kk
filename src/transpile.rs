use crate::ast::*;

pub fn transpile_statements(statements: Vec<Statement>) -> String {
    statements.into_iter().map(transpile_statement).collect()
}

pub fn transpile_statement(statement: Statement) -> String {
    match statement {
        Statement::TypeAlias { .. } => "".to_string(),
        Statement::Let { left, right, .. } => match right {
            Expression::Let { .. } => {
                return format!(
                    "const {} = (()=>{{{}}})()",
                    left.representation,
                    transpile_expression(right)
                )
            }
            _ => {
                return format!(
                    "const {} = {}",
                    left.representation,
                    transpile_expression(right)
                )
            }
        },
    }
}

fn transpile_tag(tag: String) -> String {
    tag.as_str()[1..].to_string()
}

pub fn transpile_expression(expression: Expression) -> String {
    match expression {
        Expression::String(s) => s.representation,
        Expression::Number(n) => n.representation,
        Expression::Variable(v) => v.representation,
        Expression::Tag { token, payload } => match payload {
            Some(payload) => format!(
                "{{$:'{}',_:{}}}",
                transpile_tag(token.representation),
                transpile_expression(payload.value)
            ),
            None => format!("{{$:'{}'}}", transpile_tag(token.representation)),
        },
        Expression::Record {
            spread,
            key_value_pairs,
            ..
        } => format!(
            "{{{}{}}}",
            match spread {
                None => "".to_string(),
                Some(expression) => format!("...({}),", transpile_expression(*expression)),
            },
            key_value_pairs
                .into_iter()
                .map(|RecordKeyValue { key, value, .. }| {
                    format!("{}: {}", key.representation, transpile_expression(value))
                })
                .collect::<Vec<String>>()
                .join(",")
        ),
        Expression::Function(function) => {
            let Function {
                first_branch,
                branches,
            } = *function;
            let number_of_args = first_branch.arguments.len();
            let arguments: Vec<String> = (0..number_of_args).map(|x| format!("_{}", x)).collect();
            let branches = vec![first_branch]
                .into_iter()
                .chain(branches.into_iter())
                .map(transpile_function_branch)
                .collect::<Vec<String>>()
                .join("\n");
            format!("({})=>{{{}}}", arguments.join(","), branches)
        }
        Expression::FunctionCall(FunctionCall {
            function,
            arguments,
        }) => format!(
            "({})({})",
            transpile_expression(*function),
            arguments
                .into_iter()
                .map(transpile_expression)
                .collect::<Vec<String>>()
                .join(",")
        ),
        Expression::Array(values) => format!(
            "[{}]",
            values
                .into_iter()
                .map(transpile_expression)
                .collect::<Vec<String>>()
                .join(",")
        ),
        Expression::Let {
            left,
            right,
            false_branch: else_return,
            true_branch: return_value,
            ..
        } => {
            let temp_placeholder = "$TEMP".to_string(); //TODO: get from symbol table to prevent name clashing
            let TranspiledDestructurePattern {
                bindings,
                conditions,
            } = transpile_function_destructure_pattern(left, temp_placeholder.clone());
            let init = format!(
                "var {} = {}",
                temp_placeholder,
                transpile_expression(*right)
            );
            let return_now = match *return_value {
                Expression::Let { .. } => "",
                _ => "return ",
            };
            let return_value = transpile_expression(*return_value);
            if conditions.is_empty() {
                format!(
                    "{};{};{}{}",
                    init,
                    bindings.join(";"),
                    return_now,
                    return_value
                )
            } else {
                let else_return = match else_return {
                    Some(expression) => format!(
                        "({})({})",
                        transpile_expression(*expression),
                        temp_placeholder
                    ),
                    None => temp_placeholder,
                };
                format!(
                    "{};if({}){{{};{}{}}}else{{return {}}}",
                    init,
                    conditions.join(" && "),
                    bindings.join(";"),
                    return_now,
                    return_value,
                    else_return
                )
            }
        }
    }
}

pub fn transpile_function_branch(function_branch: FunctionBranch) -> String {
    let transpiled_destructure_pattern = function_branch
        .arguments
        .into_iter()
        .enumerate()
        .map(|(index, argument)| {
            transpile_function_destructure_pattern(
                argument.destructure_pattern,
                format!("_{}", index),
            )
        })
        .fold(
            TranspiledDestructurePattern {
                conditions: vec![],
                bindings: vec![],
            },
            join_transpiled_destructure_pattern,
        );

    let return_now = match *function_branch.body {
        Expression::Let { .. } => "",
        _ => "return ",
    };
    let body = transpile_expression(*function_branch.body);
    let conditions = {
        if transpiled_destructure_pattern.conditions.is_empty() {
            "".to_string()
        } else {
            format!(
                "if({})",
                transpiled_destructure_pattern.conditions.join(" && "),
            )
        }
    };
    format!(
        "{}{{{};{}{}}}",
        conditions,
        transpiled_destructure_pattern.bindings.join(";"),
        return_now,
        body
    )
}

#[derive(Debug)]
pub struct TranspiledDestructurePattern {
    conditions: Vec<String>,
    bindings: Vec<String>,
}

pub fn transpile_function_destructure_pattern(
    destructure_pattern: DestructurePattern,
    from_expression: String,
) -> TranspiledDestructurePattern {
    match destructure_pattern {
        DestructurePattern::String(token) => TranspiledDestructurePattern {
            conditions: vec![format!("{} === {}", token.representation, from_expression)],
            bindings: vec![],
        },
        DestructurePattern::Number(token) => TranspiledDestructurePattern {
            conditions: vec![format!("{} === {}", token.representation, from_expression)],
            bindings: vec![],
        },
        DestructurePattern::Underscore(_) => TranspiledDestructurePattern {
            conditions: vec![],
            bindings: vec![],
        },
        DestructurePattern::Identifier(identifier) => TranspiledDestructurePattern {
            conditions: vec![],
            bindings: vec![format!(
                "var {} = {}",
                identifier.representation, from_expression
            )],
        },
        DestructurePattern::Tag { token, payload } => {
            let first = TranspiledDestructurePattern {
                conditions: vec![format!(
                    "{}.$==='{}'",
                    from_expression,
                    transpile_tag(token.representation)
                )],
                bindings: vec![],
            };
            let rest = match payload {
                None => None,
                Some(payload) => Some(transpile_function_destructure_pattern(
                    payload.destructure_pattern,
                    format!("{}._", from_expression),
                )),
            };
            match rest {
                None => first,
                Some(rest) => join_transpiled_destructure_pattern(first, rest),
            }
        }
        DestructurePattern::Record {
            key_value_pairs, ..
        } => key_value_pairs.into_iter().fold(
            TranspiledDestructurePattern {
                bindings: vec![],
                conditions: vec![],
            },
            |result, DestructuredRecordKeyValue { key, as_value, .. }| {
                join_transpiled_destructure_pattern(
                    result,
                    match as_value {
                        None => TranspiledDestructurePattern {
                            bindings: vec![format!(
                                "var {} = {}.{}",
                                key.representation, from_expression, key.representation
                            )],
                            conditions: vec![],
                        },
                        Some(destructure_pattern) => transpile_function_destructure_pattern(
                            destructure_pattern,
                            format!("{}.{}", from_expression, key.representation),
                        ),
                    },
                )
            },
        ),
    }
}

fn join_transpiled_destructure_pattern(
    a: TranspiledDestructurePattern,
    b: TranspiledDestructurePattern,
) -> TranspiledDestructurePattern {
    TranspiledDestructurePattern {
        conditions: a
            .conditions
            .into_iter()
            .chain(b.conditions.into_iter())
            .collect(),
        bindings: a
            .bindings
            .into_iter()
            .chain(b.bindings.into_iter())
            .collect(),
    }
}
