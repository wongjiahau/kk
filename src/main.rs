pub mod ast;
use ast::*;

mod transpile;
use transpile::*;

mod parse;
use parse::*;

mod tokenize;
use tokenize::*;

fn main() {
    println!(
        "{:?}",
        transpile_source("let f = (x: number, y) => 'helo' | x => '2'".to_string())
    )
}

pub fn transpile_source(source: String) -> Result<String, ParseError> {
    let tokens = tokenize(source)?;
    let statements = parse_statements(tokens)?;
    // println!("{:#?}", statements);
    Ok(transpile_statements(statements))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_tokenize_arrow_right() {
        assert_eq!(
            tokenize("=>".to_string()),
            Ok(vec![Token {
                token_type: TokenType::ArrowRight,
                representation: "=>".to_string(),
                position: Position {
                    column_start: 0,
                    column_end: 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        );
    }

    #[test]
    fn test_tokenize_tag() {
        assert_eq!(
            tokenize("#Hello_world123 ".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::Tag("#Hello_world123".to_string()),
                    representation: "#Hello_world123".to_string(),
                    position: Position {
                        column_start: 0,
                        column_end: 14,
                        line_start: 0,
                        line_end: 0
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 15,
                        column_end: 15,
                    }
                }
            ])
        );
    }

    #[test]
    fn test_tokenize_string() {
        let string = "'qwertyuiop \"world\" 1234567890!'".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::String(string.clone()),
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_identifier() {
        let string = "Hello_world_123".to_string();
        assert_eq!(
            tokenize(string.clone()),
            Ok(vec![Token {
                token_type: TokenType::Identifier(string.clone()),
                representation: string.clone(),
                position: Position {
                    column_start: 0,
                    column_end: string.len() - 1,
                    line_start: 0,
                    line_end: 0
                }
            }])
        )
    }

    #[test]
    fn test_tokenize_keyword() {
        assert_eq!(
            tokenize("let type".to_string()),
            Ok(vec![
                Token {
                    token_type: TokenType::KeywordLet,
                    representation: "let".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 0,
                        column_end: 2
                    }
                },
                Token {
                    token_type: TokenType::Whitespace,
                    representation: " ".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 3,
                        column_end: 3
                    }
                },
                Token {
                    token_type: TokenType::KeywordType,
                    representation: "type".to_string(),
                    position: Position {
                        line_start: 0,
                        line_end: 0,
                        column_start: 4,
                        column_end: 7
                    }
                }
            ])
        )
    }

    #[test]
    fn test_let_statement_1() {
        assert_eq!(
            transpile_source("let x = '123'".to_string()),
            Ok("const x = '123'".to_string())
        )
    }

    #[test]
    fn test_let_statement_2() {
        assert_eq!(
            transpile_source("let x = y".to_string()),
            Ok("const x = y".to_string())
        )
    }

    #[test]
    fn test_function_1() {
        assert_debug_snapshot!(transpile_source("let x = \\x => 'yo'".to_string()))
    }
    #[test]
    fn test_function_2() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#red => 'red' \\ #blue => 'blue'".to_string()
        ))
    }

    #[test]
    fn test_function_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = \\#ok(#some(r)) => r \\ _ => 'yo'".to_string()
        ))
    }

    #[test]
    fn test_function_4() {
        assert_debug_snapshot!(transpile_source(
            "let or = \\#false, #false => #false \\ _, _ => #true".to_string()
        ))
    }

    #[test]
    fn test_function_5() {
        assert_debug_snapshot!(transpile_source("let square = \\x:number => x".to_string()))
    }

    #[test]
    fn test_function_call_1() {
        assert_debug_snapshot!(transpile_source("let x = y.square.bomb".to_string()))
    }

    #[test]
    fn test_function_call_2() {
        assert_debug_snapshot!(transpile_source("let x = a.times(b).and(c, d)".to_string()))
    }

    #[test]
    fn test_function_call_3() {
        assert_debug_snapshot!(transpile_source(
            "let x = x.(\\#ok(x) => a \\#no => b)".to_string()
        ))
    }

    #[test]
    fn test_function_call_4() {
        assert_debug_snapshot!(transpile_source(
            "let x = 'hello world'.replace('hello', with: 'bye')".to_string()
        ))
    }
}
