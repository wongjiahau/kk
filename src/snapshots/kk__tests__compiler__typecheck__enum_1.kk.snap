---
source: src/main.rs
expression: stripped_actual_output.trim()
---
============
INPUT
============
enum List<Element> =
  #nil
  #cons({
    current: Element,
    next: List<Element= Element>
  })

let xs = #cons({
  current = 1,
  next = #cons({
    current = true,
    next = #nil
  })
})

============
EXIT CODE
============
1

============
STDOUT
============


============
STDERR
============
error:
   ┌─ tests/compiler/typecheck/enum_1.kk:8:16
   │
 8 │   let xs = #cons({
   │ ╭────────────────^
 9 │ │   current = 1,
10 │ │   next = #cons({
11 │ │     current = true,
12 │ │     next = #nil
13 │ │   })
14 │ │ })
   │ ╰─^ Type mismatch
   │
   = +-------------------+--------------------+
     |   EXPECTED TYPE   |    ACTUAL TYPE     |
     +-------------------+--------------------+
     |   {               |   {                |
     |     current:      |     current:       |
     |       number,     |       number,      |
     |     next:         |     next:          |
     |       List<       |       List<        |
     |         Element=  |         Element=   |
     | -         number, | +         boolean, |
     |       >,          |       >,           |
     |   }               |   }                |
     +-------------------+--------------------+
